Sources/Alchemy/Database/QueryBuilder/Grammar.swift

```diff
@@ -0,0 +1,240 @@
+import Foundation
+
+public class Grammar {
+
+    enum GrammarError: Error {
+        case missingTable
+    }
+
+    private let selectComponents: [AnyKeyPath] = [
+        \Query.columns,
+        \Query.from,
+        \Query.joins,
+        \Query.wheres,
+        \Query.groups,
+        \Query.havings,
+        \Query.orders,
+        \Query.limit,
+        \Query.offset
+    ]
+
+    func compileSelect(query: Query) -> SQL {
+
+        // If the query does not have any columns set, we'll set the columns to the
+        // * character to just get all of the columns from the database. Then we
+        // can build the query and concatenate all the pieces together as one.
+        let original = query.columns
+
+        if query.columns == nil {
+            query.columns = [SQL("*")]
+        }
+
+        // To compile the query, we'll spin through each component of the query and
+        // see if that component exists. If it does we'll just call the compiler
+        // function for the component which is responsible for making the SQL.
+        let sql = compileComponents(query: query)
+
+        query.columns = original
+
+        return sql
+    }
+
+    private func compileComponents(query: Query) -> SQL {
+        var sql: [String] = []
+        var bindings: [DatabaseValue] = []
+
+        for component in selectComponents {
+            if let part = query[keyPath: component] {
+                if component == \Query.columns,
+                    let columns = part as? [Raw] {
+                    compileColumns(query, columns: columns)
+                        .bind(queries: &sql, bindings: &bindings)
+                }
+                else if component == \Query.from,
+                    let table = part as? String {
+                    sql.append(compileFrom(query, table: table))
+                }
+                else if component == \Query.joins,
+                    let joins = part as? [JoinClause] {
```

Comment:
Could do inversion of control here to avoid all the as? casts; i.e. each clause has a function

`func compile(with: Grammar) {}`